diff --git a/usr/lib/ocf/resource.d/heartbeat/mysql b/mysql-wss
index 43d155d..a655ce1 100755
--- a/usr/lib/ocf/resource.d/heartbeat/mysql
+++ b/usr/lib/ocf/resource.d/heartbeat/mysql
@@ -746,6 +746,41 @@ get_reader_attr() {
 
 }
 
+rebuild_cluster() {
+    if [ "$(crm_attribute -t crm_config --name mysqlprimaryinit --query 2> /dev/null | awk -F\= '{print $4}')" = 'done' ] ; then
+        mysql_status info()
+        if [ $? = $OCF_SUCCESS ] ; then
+            CRM_ONLINE_NODES=( $(crm node list | sed 's/\://g') )
+            WSREP_EPOCH=$(echo 'show status;' | mysql -u root | grep wsrep_protocol_version | awk '{print $2}')
+            WSREP_STATE=$(echo 'show status;' | mysql -u root | grep wsrep_cluster_status | awk '{print $2}')
+            if [ "$WSREP_STATE" = 'non-Primary' ]; then
+                WINNER_NODE=$(hostname)
+                MAX_EPOCH=$WSREP_EPOCH
+                crm_attribute -t crm_config --node $(hostname) --name mysqlepoch --update $WSREP_EPOCH
+                for node in "${CRM_ONLINE_NODES[@]}" ; do
+                    CUR_EPOCH=$(crm_attribute -t crm_config --node $node --name mysqlepoch --query 2> /dev/null | awk -F\= '{print $4}')
+                    if [ -z "$CUR_EPOCH" ] ; then
+                      echo "Skipping comparison with node $node"
+                    else
+                        if [ "$CUR_EPOCH" -gt "$MAX_EPOCH" ] ; then
+                          WINNER_NODE=$node
+                          MAX_EPOCH=$CUR_EPOCH
+                        fi
+                    fi
+                done
+                echo $WINNER_NODE
+                if [ "$WINNER_NODE" = "$(hostname)" ] ; then
+                   ocf_log info "Assuming current node as a new master..."
+                   /etc/init.d/mysql stop
+                   /etc/init.d/mysql start --wsrep-cluster-address="gcomm://"
+                fi
+            else
+                return $OCF_SUCCESS
+            fi
+        fi
+    fi
+}
+
 # Stores data for MASTER STATUS from MySQL
 update_data_master_status() {
 
@@ -1019,6 +1054,8 @@ mysql_start() {
     fi
 
     ocf_log info "MySQL started"
+    sleep 10
+    rebuild_cluster
     return $OCF_SUCCESS
 }
 
